### ISSUE TITLE
Translating LambdaExpression Compile fails in PCL

### ISSUE THREAD

**Issue by dmirmilshteyn** [sq/JSIL#758]
I'm trying to compile an expression tree in a PCL, and am running into the following error. It only happens when compiled as a PCL though; a regular class library works fine.

```
System.AggregateException: One or more errors occurred. ---> System.Exception: Error occurred while generating javascript for assembly 'PCLTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. ---> System.Exception: An error occurred while declaring the type 'PCLTest.Class1' ---> System.Exception: Error occurred while generating javascript for method 'System.Void PCLTest.Class1::Main()'. ---> System.NotImplementedException: Unimplemented form of generic type parameter: 'TDelegate'.
   at JSIL.Internal.JavascriptFormatter.TypeReferenceInternal(GenericParameter gp, TypeReferenceContext context) in C:\Projects\JSIL\JSIL\JavascriptFormatter.cs:line 550
   at JSIL.Internal.JavascriptFormatter.TypeReference(TypeReference type, TypeReferenceContext context) in C:\Projects\JSIL\JSIL\JavascriptFormatter.cs:line 733
   at JSIL.Internal.JavascriptFormatter.Signature(MethodReference method, MethodSignature signature, TypeReferenceContext context, Boolean forConstructor, Boolean allowCache) in C:\Projects\JSIL\JSIL\JavascriptFormatter.cs:line 1242
   at JSIL.JavascriptAstEmitter.VisitNode(JSLocalCachedSignatureExpression lcse) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 2450
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.VisitAndMaybeHintLoad(JSExpression value, Boolean targetIsLocal) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 1829
   at JSIL.JavascriptAstEmitter.VisitNode(JSBinaryOperatorExpression bop) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 1910
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.CommaSeparatedList(IEnumerable`1 values, Boolean withNewlines) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 72
   at JSIL.JavascriptAstEmitter.VisitNode(JSVariableDeclarationStatement vars) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 309
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.VisitNode(JSBlockStatement block, Boolean includeBraces) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 129
   at JSIL.JavascriptAstEmitter.VisitNode(JSBlockStatement block) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 117
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.VisitNode(JSFunctionExpression function) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 1404
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.AssemblyTranslator.EmitMethodBody(DecompilerContext context, MethodReference methodRef, MethodDefinition method, JavascriptAstEmitter astEmitter, JavascriptFormatter output, Boolean stubbed, Cachers cachers, Int32& nextDisambiguatedId, MethodInfo methodInfo, Action`1 bodyTransformer) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 2825
   --- End of inner exception stack trace ---
   at JSIL.AssemblyTranslator.EmitMethodBody(DecompilerContext context, MethodReference methodRef, MethodDefinition method, JavascriptAstEmitter astEmitter, JavascriptFormatter output, Boolean stubbed, Cachers cachers, Int32& nextDisambiguatedId, MethodInfo methodInfo, Action`1 bodyTransformer) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 2827
   at JSIL.AssemblyTranslator.EmitTypeMethodExpressions(DecompilerContext context, TypeDefinition typedef, JavascriptAstEmitter astEmitter, JavascriptFormatter output, Boolean stubbed, JSRawOutputIdentifier dollar, Boolean makingSkeletons, Int32& nextDisambiguatedId) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 1759
   at JSIL.AssemblyTranslator.DeclareType(DecompilerContext context, TypeDefinition typedef, JavascriptAstEmitter astEmitter, JavascriptFormatter output, HashSet`1 declaredTypes, Boolean stubbed) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 1384
   --- End of inner exception stack trace ---
   at JSIL.AssemblyTranslator.DeclareType(DecompilerContext context, TypeDefinition typedef, JavascriptAstEmitter astEmitter, JavascriptFormatter output, HashSet`1 declaredTypes, Boolean stubbed) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 1560
   at JSIL.AssemblyTranslator.TranslateModule(DecompilerContext context, JavascriptFormatter output, ModuleDefinition module, HashSet`1 sealedTypes, HashSet`1 declaredTypes, Boolean stubbed) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 986
   at JSIL.AssemblyTranslator.TranslateSingleAssemblyInternal(DecompilerContext context, AssemblyDefinition assembly, Stream outputStream) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 921
   at JSIL.AssemblyTranslator.<>c__DisplayClass83_0.<TranslateInternal>b__0(Int32 i) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 525
   --- End of inner exception stack trace ---
   at JSIL.AssemblyTranslator.<>c__DisplayClass83_0.<TranslateInternal>b__0(Int32 i) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 527
   at System.Threading.Tasks.Parallel.<>c__DisplayClass9`1.<ForWorker>b__11()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass0.<ExecuteSelfReplicating>b__1(Object )
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
   at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body)
   at JSIL.AssemblyTranslator.TranslateInternal(String assemblyPath, Boolean scanForProxies) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 551
   at JSIL.AssemblyTranslator.Translate(String assemblyPath, Boolean scanForProxies) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 443
   at JSIL.Compiler.Profiles.Default.Translate(VariableSet variables, AssemblyTranslator translator, Configuration configuration, String assemblyPath, Boolean scanForProxies) in C:\Projects\JSIL\Compiler\Profiles\Default.cs:line 23
   at JSIL.Compiler.Program.InternalMain(String[] arguments) in C:\Projects\JSIL\Compiler\Program.cs:line 752
   at JSIL.Compiler.Program.Main(String[] arguments) in C:\Projects\JSIL\Compiler\Program.cs:line 631
---> (Inner Exception #0) System.Exception: Error occurred while generating javascript for assembly 'PCLTest, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'. ---> System.Exception: An error occurred while declaring the type 'PCLTest.Class1' ---> System.Exception: Error occurred while generating javascript for method 'System.Void PCLTest.Class1::Main()'. ---> System.NotImplementedException: Unimplemented form of generic type parameter: 'TDelegate'.
   at JSIL.Internal.JavascriptFormatter.TypeReferenceInternal(GenericParameter gp, TypeReferenceContext context) in C:\Projects\JSIL\JSIL\JavascriptFormatter.cs:line 550
   at JSIL.Internal.JavascriptFormatter.TypeReference(TypeReference type, TypeReferenceContext context) in C:\Projects\JSIL\JSIL\JavascriptFormatter.cs:line 733
   at JSIL.Internal.JavascriptFormatter.Signature(MethodReference method, MethodSignature signature, TypeReferenceContext context, Boolean forConstructor, Boolean allowCache) in C:\Projects\JSIL\JSIL\JavascriptFormatter.cs:line 1242
   at JSIL.JavascriptAstEmitter.VisitNode(JSLocalCachedSignatureExpression lcse) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 2450
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.VisitAndMaybeHintLoad(JSExpression value, Boolean targetIsLocal) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 1829
   at JSIL.JavascriptAstEmitter.VisitNode(JSBinaryOperatorExpression bop) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 1910
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.CommaSeparatedList(IEnumerable`1 values, Boolean withNewlines) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 72
   at JSIL.JavascriptAstEmitter.VisitNode(JSVariableDeclarationStatement vars) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 309
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.VisitNode(JSBlockStatement block, Boolean includeBraces) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 129
   at JSIL.JavascriptAstEmitter.VisitNode(JSBlockStatement block) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 117
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.JavascriptAstEmitter.VisitNode(JSFunctionExpression function) in C:\Projects\JSIL\JSIL\JavascriptAstEmitter.cs:line 1404
   at JSIL.Ast.JSAstVisitor.VisitorCache.Adapter`2.Visit(JSAstVisitor this, JSNode node) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 58
   at JSIL.Ast.JSAstVisitor.Visit(JSNode node, String name) in C:\Projects\JSIL\JSIL\AST\JSAstVisitor.cs:line 203
   at JSIL.AssemblyTranslator.EmitMethodBody(DecompilerContext context, MethodReference methodRef, MethodDefinition method, JavascriptAstEmitter astEmitter, JavascriptFormatter output, Boolean stubbed, Cachers cachers, Int32& nextDisambiguatedId, MethodInfo methodInfo, Action`1 bodyTransformer) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 2825
   --- End of inner exception stack trace ---
   at JSIL.AssemblyTranslator.EmitMethodBody(DecompilerContext context, MethodReference methodRef, MethodDefinition method, JavascriptAstEmitter astEmitter, JavascriptFormatter output, Boolean stubbed, Cachers cachers, Int32& nextDisambiguatedId, MethodInfo methodInfo, Action`1 bodyTransformer) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 2827
   at JSIL.AssemblyTranslator.EmitTypeMethodExpressions(DecompilerContext context, TypeDefinition typedef, JavascriptAstEmitter astEmitter, JavascriptFormatter output, Boolean stubbed, JSRawOutputIdentifier dollar, Boolean makingSkeletons, Int32& nextDisambiguatedId) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 1759
   at JSIL.AssemblyTranslator.DeclareType(DecompilerContext context, TypeDefinition typedef, JavascriptAstEmitter astEmitter, JavascriptFormatter output, HashSet`1 declaredTypes, Boolean stubbed) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 1384
   --- End of inner exception stack trace ---
   at JSIL.AssemblyTranslator.DeclareType(DecompilerContext context, TypeDefinition typedef, JavascriptAstEmitter astEmitter, JavascriptFormatter output, HashSet`1 declaredTypes, Boolean stubbed) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 1560
   at JSIL.AssemblyTranslator.TranslateModule(DecompilerContext context, JavascriptFormatter output, ModuleDefinition module, HashSet`1 sealedTypes, HashSet`1 declaredTypes, Boolean stubbed) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 986
   at JSIL.AssemblyTranslator.TranslateSingleAssemblyInternal(DecompilerContext context, AssemblyDefinition assembly, Stream outputStream) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 921
   at JSIL.AssemblyTranslator.<>c__DisplayClass83_0.<TranslateInternal>b__0(Int32 i) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 525
   --- End of inner exception stack trace ---
   at JSIL.AssemblyTranslator.<>c__DisplayClass83_0.<TranslateInternal>b__0(Int32 i) in C:\Projects\JSIL\JSIL\AssemblyTranslator.cs:line 527
   at System.Threading.Tasks.Parallel.<>c__DisplayClass9`1.<ForWorker>b__11()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(Task childTask)
   at System.Threading.Tasks.Task.<>c__DisplayClass0.<ExecuteSelfReplicating>b__1(Object )<---
```

Here's a test case. Will fail during the "Writing JS" phase when compiled with PCL profile 151.

```csharp
        public static void Test() {
            ParameterExpression numParam = Expression.Parameter(typeof(int));
            ConstantExpression five = Expression.Constant(5, typeof(int));
            Expression numLessThanFive = Expression.LessThan(numParam, five);
            Expression<Func<int, bool>> lambda1 = Expression.Lambda<Func<int, bool>>(numLessThanFive, new ParameterExpression[] { numParam });

            Func<int, bool> func = lambda1.Compile();
        }
```

**Comment by iskiselev** [comment]
Very interesting issue - I believe it could be fixed very simple in this particular case, but may be tricky in others.

When we build using PCL profile 151, we end with comparing (don't know why yet) of `Expression<T>` from `System.Core` and `Expression<T>` from `System.Linq.Expressions`. In run-time last assembly should be replaced with assembly that will redirect `Expression<T>` back to `System.Core` or other assembly depending on runtime. They are not equal for JSIL.

Here probem could be solved by adding to `.jsilconfig` Assemblies section something like:
```
"Redirects" :
    {
      "System.Linq.Expressions" : "System.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089"
    }
```

But what should we do if classes are really redirected to different assemblies (some to  `System.Core`, other to `mscorlib` for example)? Should we load provided with .Net Framework runtime redirection? And what we do if Desktop framework is not in supported list of profile?

**Comment by iskiselev** [comment]
Probably this is to late - but I see a potential problem in `MethodSignature.ResolveGenericParameter`. It tries to compare types using `TypeUtil.TypesAreEqual(gpOwner, owner, false)` - and they are different, as they are defined in different assemblies. but if we do `TypeUtil.TypesAreEqual(gpOwner.Resolve(), owner.Resolve, false)` - they will be same, as TypeRedirection is on GAC version of `System.Linq.Expressions`.

Looks like somewhere we have done `Resolve` to get TypeDefenition from TypeReference - otherwise I'd suppose to have both parameters from `System.Core`.

**Comment by iskiselev** [comment]
Attempt to fix it could be looked in ff3aea706c7ef652d042a3092c07e262bee0e3c0, but it breaks some tests. Will investigate some more next days - or probably @kg has some ideas.

**Comment by iskiselev** [comment]
Better fix is 2f3c76ba900e7a91b7b06395c1bf422cf7a558dd, but I still need to think a little bit more on it before preparing PR. Looks like this version of fix passes all tests.

**Comment by dmirmilshteyn** [comment]
Awesome, that worked great for me. Thanks for the quick fix!
